src/jupyterlabcontroller/storage/kubernetes/custom.py

```diff
@@ -0,0 +1,303 @@
+        namespace
+            Namespace of the object.
+        wait
+            Whether to wait for the custom object to be deleted.
+
+        Raises
+        ------
+        KubernetesError
+            Raised for exceptions from the Kubernetes API server.
+        """
+        msg = f"Deleting {self._kind}"
+        self._logger.debug(msg, name=name, namespace=namespace)
+        try:
+            await self._api.delete_namespaced_custom_object(
+                self._group, self._version, namespace, self._plural, name
+            )
+        except ApiException as e:
+            if e.status == 404:
+                return
+            raise KubernetesError.from_exception(
+                "Error deleting object",
+                e,
+                kind=self._kind,
+                namespace=namespace,
+                name=name,
+            ) from e
+        if wait:
+            await self.wait_for_deletion(name, namespace)
+
+    async def list(self, namespace: str) -> list[dict[str, Any]]:
+        """List the custom objects in a namespace.
+
+        Parameters
+        ----------
+        namespace
+            Namespace in which to list custom objects.
+
+        Returns
+        -------
+        list of dict
+            List of custom objects found.
+
+        Raises
+        ------
+        KubernetesError
+            Raised for exceptions from the Kubernetes API server.
+        """
+        try:
+            objs = await self._api.list_namespaced_custom_object(
+                self._group, self._version, namespace, self._plural
+            )
+        except ApiException as e:
+            raise KubernetesError.from_exception(
+                "Error listing objects",
+                e,
+                kind=self._kind,
+                namespace=namespace,
+            ) from e
+        return objs.items
+
+    async def read(self, name: str, namespace: str) -> dict[str, Any] | None:
+        """Read a custom object.
+
+        Parameters
+        ----------
+        name
+            Name of the custom object.
+        namespace
+            Namespace of the custom object.
+
+        Returns
+        -------
+        dict or None
+            Custom object, or `None` if it does not exist.
+
+        Raises
+        ------
+        KubernetesError
+            Raised for exceptions from the Kubernetes API server.
+        """
+        try:
+            return await self._api.get_namespaced_custom_object(
+                self._group, self._version, namespace, self._plural, name
+            )
+        except ApiException as e:
+            if e.status == 404:
+                return None
+            raise KubernetesError.from_exception(
+                "Error reading object",
+                e,
+                kind=self._kind,
+                namespace=namespace,
+                name=name,
+            ) from e
+
+    async def wait_for_deletion(self, name: str, namespace: str) -> None:
```

Comment:
I don't know why wait_for_deletion bugs me more than the rest of the duplicated stuff.  Maybe it's that watch logic that seems to be repeated almost verbatim.

Reply:
Yeah, I think I might be able to move some of the logic into the watcher class, although not the construction of the watcher object itself. I'll do that in an upcoming PR.